Stabilize `anonymous_pipe`

tbu- opened the pull request:
Closes #127154.

Based on top of #135635 to avoid conflicts.

tbu- committed:
Move `std::io::pipe` code into its own file

tbu- committed:
Update `std::io::{pipe, PipeReader, PipeWriter}` docs the new location

Also create a section "Platform-specific behavior", don't hide required
imports for code examples.

tbu- committed:
pipe: Remove unnecessary miri exclusions from no_run code

tbu- committed:
Stabilize `anonymous_pipe`

Closes #127154.

tbu- commented:
Diff for the two commits unique to this PR: https://github.com/rust-lang/rust/pull/135822/files/220b34a987ce5dd4213803470caccb05f8a25173..42584e17db0e8979e31e2fe1a7c5a14ec33f559c.

tbu- commented:
CC @NobodyXu

NobodyXu reviewed (approved):
Thank you!


I'm so happy that this is getting stabilised 🎉

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12888187247/job/35932516624) [(plain)](https://github.com/rust-lang/rust/commit/42584e17db0e8979e31e2fe1a7c5a14ec33f559c/checks/35932516624/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- library/std/src/io/pipe.rs - io::pipe::PipeWriter::try_clone (line 135) stdout ----
error[E0282]: type annotations needed
  --> library/std/src/io/pipe.rs:158:1
   |
26 | Ok(())
   | ^^ cannot infer type of the type parameter `E` declared on the enum `Result`
help: consider specifying the generic arguments
   |
   |
26 | Ok::<(), E>(())

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0282`.
For more information about this error, try `rustc --explain E0282`.
Couldn't compile the test.
---- library/std/src/io/pipe.rs - io::pipe::PipeReader::try_clone (line 79) stdout ----
error[E0282]: type annotations needed
  --> library/std/src/io/pipe.rs:122:1
   |
46 | Ok(())
   | ^^ cannot infer type of the type parameter `E` declared on the enum `Result`
help: consider specifying the generic arguments
   |
   |
46 | Ok::<(), E>(())

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0282`.
For more information about this error, try `rustc --explain E0282`.
Couldn't compile the test.
---- library/std/src/io/pipe.rs - io::pipe::pipe (line 34) stdout ----
error[E0282]: type annotations needed
  --> library/std/src/io/pipe.rs:54:1
   |
23 | Ok(())
   | ^^ cannot infer type of the type parameter `E` declared on the enum `Result`
help: consider specifying the generic arguments
   |
   |
23 | Ok::<(), E>(())

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0282`.
---
    library/std/src/io/pipe.rs - io::pipe::pipe (line 34)

test result: FAILED. 1307 passed; 3 failed; 19 ignored; 0 measured; 0 filtered out; finished in 43.83s

error: doctest failed, to rerun pass `-p std --doc`
  local time: Tue Jan 21 14:04:20 UTC 2025
  network time: Tue, 21 Jan 2025 14:04:20 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12947793888/job/36115112942) [(plain)](https://github.com/rust-lang/rust/commit/aba3017c29fedf35faa5b6e0bdd493de49768e4b/checks/36115112942/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tests/pass/float_nan.rs ... ok
tests/pass/0weak_memory_consistency_sc.rs ... ok

FAILED TEST: tests/pass/shims/pipe.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-NqA3Qg" RUST_BACKTRACE="1" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/miri" "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/miri_ui/tests/pass/shims" "tests/pass/shims/pipe.rs" "--edition" "2021"
error: test got exit status: 1, but expected 0
 = note: compilation failed, but was expected to succeed

error: actual output differed from expected
---

Location:
   /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ui_test-0.26.5/src/lib.rs:357

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
error: test failed, to rerun pass `--test ui`
Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-c9642fae2e8182ec --quiet` (exit status: 1)
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-c9642fae2e8182ec --quiet` (exit status: 1)
Command has failed. Rerun with -v to see more details.
  local time: Fri Jan 24 10:56:24 UTC 2025
  network time: Fri, 24 Jan 2025 10:56:24 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

joboet commented:
The changes here look good to me, though I'd like to see another change before stabilization: 

Currently, things like the `Stdio: From<PipeWriter>` implementations live in `sys::` which IMHO is really confusing and dangerous (because people will forget to add these implementations when porting `std`). Could you move all the public trait implementations to `io::pipe` (and `os::_::pipe` for the platform-specific implementations)?

joboet reviewed (changes requested):
The changes here look good to me, though I'd like to see another change before stabilization: 

Currently, things like the `Stdio: From<PipeWriter>` implementations live in `sys::` which IMHO is really confusing and dangerous (because people will forget to add these implementations when porting `std`). Could you move all the public trait implementations to `io::pipe` (and `os::_::pipe` for the platform-specific implementations)?

bors commented:
:umbrella: The latest upstream changes (presumably #136070) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tbu- committed:
Move trait impls for `Pipe{Reader,Writer}` to `std::io::pipe`

Also get rid of the now mostly-empty `std::sys::anonymous_pipe` module.
It basically just unwrapped the `AnonPipe` abstraction from
`std::sys::pipe`.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12975127560/job/36185809560) [(plain)](https://github.com/rust-lang/rust/commit/58154046a31c9af0ea7cd3131e31f7bbb470bb74/checks/36185809560/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking panic_abort v0.0.0 (/checkout/library/panic_abort)
    Checking panic_unwind v0.0.0 (/checkout/library/panic_unwind)
    Checking std_detect v0.1.5 (/checkout/library/stdarch/crates/std_detect)
    Checking hashbrown v0.15.2
error[E0425]: cannot find value `owned` in this scope
   --> library/std/src/os/windows/io/handle.rs:564:80
    |
564 |         io::PipeReader::from_inner(FromInner::from_inner(FromInner::from_inner(owned)))

error[E0412]: cannot find type `PipeWriter` in this scope
error[E0412]: cannot find type `PipeWriter` in this scope
   --> library/std/src/os/windows/io/handle.rs:569:19
569 | impl AsHandle for PipeWriter {
    |                   ^^^^^^^^^^ not found in this scope
    |
help: consider importing this struct through its public re-export
help: consider importing this struct through its public re-export
    |
5   + use crate::io::PipeWriter;
    |

error[E0412]: cannot find type `PipeWriter` in this scope
   --> library/std/src/os/windows/io/handle.rs:575:11
    |
575 | impl From<PipeWriter> for OwnedHandle {
    |
help: consider importing this struct through its public re-export
    |
5   + use crate::io::PipeWriter;
5   + use crate::io::PipeWriter;
    |

error[E0412]: cannot find type `PipeWriter` in this scope
   --> library/std/src/os/windows/io/handle.rs:576:19
    |
576 |     fn from(pipe: PipeWriter) -> OwnedHandle {
    |
help: consider importing this struct through its public re-export
    |
5   + use crate::io::PipeWriter;
5   + use crate::io::PipeWriter;
    |

error[E0425]: cannot find value `owned` in this scope
   --> library/std/src/os/windows/io/handle.rs:583:80
    |
583 |         io::PipeWriter::from_inner(FromInner::from_inner(FromInner::from_inner(owned)))

error[E0412]: cannot find type `PipeReader` in this scope
error[E0412]: cannot find type `PipeReader` in this scope
   --> library/std/src/os/windows/io/handle.rs:588:22
588 | impl AsRawHandle for PipeReader {
    |                      ^^^^^^^^^^ not found in this scope
    |
help: consider importing this struct through its public re-export
help: consider importing this struct through its public re-export
    |
5   + use crate::io::PipeReader;
    |

error[E0412]: cannot find type `PipeReader` in this scope
   --> library/std/src/os/windows/io/handle.rs:595:24
595 | impl FromRawHandle for PipeReader {
    |                        ^^^^^^^^^^ not found in this scope
    |
help: consider importing this struct through its public re-export
help: consider importing this struct through its public re-export
    |
5   + use crate::io::PipeReader;
    |

error[E0433]: failed to resolve: use of undeclared type `Handle`
   --> library/std/src/os/windows/io/handle.rs:597:23
    |
597 |         unsafe { Self(Handle::from_raw_handle(raw_handle)) }
    |
help: a trait with a similar name exists
    |
    |
597 |         unsafe { Self(AsHandle::from_raw_handle(raw_handle)) }
help: consider importing this struct
    |
5   + use crate::sys::handle::Handle;
    |
    |

error[E0412]: cannot find type `PipeReader` in this scope
   --> library/std/src/os/windows/io/handle.rs:601:24
601 | impl IntoRawHandle for PipeReader {
    |                        ^^^^^^^^^^ not found in this scope
    |
help: consider importing this struct through its public re-export
help: consider importing this struct through its public re-export
    |
5   + use crate::io::PipeReader;
    |

error[E0412]: cannot find type `PipeWriter` in this scope
   --> library/std/src/os/windows/io/handle.rs:608:22
608 | impl AsRawHandle for PipeWriter {
    |                      ^^^^^^^^^^ not found in this scope
    |
help: consider importing this struct through its public re-export
help: consider importing this struct through its public re-export
    |
5   + use crate::io::PipeWriter;
    |

error[E0412]: cannot find type `PipeWriter` in this scope
   --> library/std/src/os/windows/io/handle.rs:615:24
615 | impl FromRawHandle for PipeWriter {
    |                        ^^^^^^^^^^ not found in this scope
    |
help: consider importing this struct through its public re-export
help: consider importing this struct through its public re-export
    |
5   + use crate::io::PipeWriter;
    |

error[E0433]: failed to resolve: use of undeclared type `Handle`
   --> library/std/src/os/windows/io/handle.rs:617:23
    |
617 |         unsafe { Self(Handle::from_raw_handle(raw_handle)) }
    |
help: a trait with a similar name exists
    |
    |
617 |         unsafe { Self(AsHandle::from_raw_handle(raw_handle)) }
help: consider importing this struct
    |
5   + use crate::sys::handle::Handle;
    |
    |

error[E0412]: cannot find type `PipeWriter` in this scope
   --> library/std/src/os/windows/io/handle.rs:621:24
621 | impl IntoRawHandle for PipeWriter {
    |                        ^^^^^^^^^^ not found in this scope
    |
help: consider importing this struct through its public re-export
---
    |
212 |             let handle = handle as sys::c::HANDLE;
    |                          ^^^^^^ not found in this scope

error[E0520]: `Error` specializes an item from a parent `impl`, but that item is not marked `default`
   --> library/std/src/os/windows/io/handle.rs:158:5
158 |     type Error = NullHandleError;
    |     ^^^^^^^^^^
    |
    = note: parent implementation is in crate `core`
    = note: parent implementation is in crate `core`

error[E0520]: `try_from` specializes an item from a parent `impl`, but that item is not marked `default`
   --> library/std/src/os/windows/io/handle.rs:161:5
    |
161 |     fn try_from(handle_or_null: HandleOrNull) -> Result<Self, NullHandleError> {
    |
    = note: parent implementation is in crate `core`

error[E0520]: `Error` specializes an item from a parent `impl`, but that item is not marked `default`
error[E0520]: `Error` specializes an item from a parent `impl`, but that item is not marked `default`
   --> library/std/src/os/windows/io/handle.rs:236:5
236 |     type Error = InvalidHandleError;
    |     ^^^^^^^^^^
    |
    = note: parent implementation is in crate `core`
    = note: parent implementation is in crate `core`

error[E0520]: `try_from` specializes an item from a parent `impl`, but that item is not marked `default`
   --> library/std/src/os/windows/io/handle.rs:239:5
    |
239 |     fn try_from(handle_or_invalid: HandleOrInvalid) -> Result<Self, InvalidHandleError> {
    |
    = note: parent implementation is in crate `core`

error[E0277]: `AnonPipe` doesn't implement `core::fmt::Debug`
error[E0277]: `AnonPipe` doesn't implement `core::fmt::Debug`
   --> library/std/src/io/pipe.rs:78:5
    |
76  | #[derive(Debug)]
    |          ----- in this derive macro expansion
77  | pub struct PipeReader {
78  |     inner: AnonPipe,
    |     ^^^^^^^^^^^^^^^ `AnonPipe` cannot be formatted using `{:?}`
   ::: /checkout/library/core/src/fmt/mod.rs:910:5
    |
910 |     pub macro Debug($item:item) {
    |     --------------- in this expansion of `#[derive(Debug)]`
    |     --------------- in this expansion of `#[derive(Debug)]`
    |
    = help: the trait `core::fmt::Debug` is not implemented for `AnonPipe`
    = note: add `#[derive(Debug)]` to `AnonPipe` or manually `impl core::fmt::Debug for AnonPipe`
help: consider annotating `AnonPipe` with `#[derive(Debug)]`
   --> library/std/src/sys/pal/windows/pipe.rs:19:1
19  + #[derive(Debug)]
20  | pub struct AnonPipe {
    |


error[E0277]: `AnonPipe` doesn't implement `core::fmt::Debug`
   --> library/std/src/io/pipe.rs:85:5
    |
83  | #[derive(Debug)]
    |          ----- in this derive macro expansion
84  | pub struct PipeWriter {
85  |     inner: AnonPipe,
    |     ^^^^^^^^^^^^^^^ `AnonPipe` cannot be formatted using `{:?}`
   ::: /checkout/library/core/src/fmt/mod.rs:910:5
    |
910 |     pub macro Debug($item:item) {
    |     --------------- in this expansion of `#[derive(Debug)]`
    |     --------------- in this expansion of `#[derive(Debug)]`
    |
    = help: the trait `core::fmt::Debug` is not implemented for `AnonPipe`
    = note: add `#[derive(Debug)]` to `AnonPipe` or manually `impl core::fmt::Debug for AnonPipe`
help: consider annotating `AnonPipe` with `#[derive(Debug)]`
   --> library/std/src/sys/pal/windows/pipe.rs:19:1
19  + #[derive(Debug)]
20  | pub struct AnonPipe {
    |


error[E0599]: the method `as_handle` exists for reference `&AnonPipe`, but its trait bounds were not satisfied
   --> library/std/src/os/windows/io/handle.rs:552:25
552 |         self.as_inner().as_handle()
    |                         ^^^^^^^^^
    |
   ::: library/std/src/sys/pal/windows/pipe.rs:19:1
   ::: library/std/src/sys/pal/windows/pipe.rs:19:1
    |
19  | pub struct AnonPipe {
    | ------------------- doesn't satisfy `AnonPipe: os::windows::io::handle::AsHandle`
note: trait bound `AnonPipe: os::windows::io::handle::AsHandle` was not satisfied
note: trait bound `AnonPipe: os::windows::io::handle::AsHandle` was not satisfied
   --> library/std/src/os/windows/io/handle.rs:445:9
    |
445 | impl<T: AsHandle + ?Sized> AsHandle for &T {
    |         |
    |         unsatisfied trait bound introduced here
note: the trait `os::windows::io::handle::AsHandle` must be implemented
note: the trait `os::windows::io::handle::AsHandle` must be implemented
   --> library/std/src/os/windows/io/handle.rs:426:1
426 | pub trait AsHandle {
    | ^^^^^^^^^^^^^^^^^^
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `os::windows::io::handle::AsHandle` defines an item `as_handle`, perhaps you need to implement it
   --> library/std/src/os/windows/io/handle.rs:426:1
426 | pub trait AsHandle {
    | ^^^^^^^^^^^^^^^^^^

error: the `Self` constructor can only be used with tuple or unit structs
error: the `Self` constructor can only be used with tuple or unit structs
   --> library/std/src/os/windows/io/handle.rs:597:18
    |
597 |         unsafe { Self(Handle::from_raw_handle(raw_handle)) }

error: the `Self` constructor can only be used with tuple or unit structs
error: the `Self` constructor can only be used with tuple or unit structs
   --> library/std/src/os/windows/io/handle.rs:617:18
    |
617 |         unsafe { Self(Handle::from_raw_handle(raw_handle)) }


error[E0599]: no method named `as_raw_handle` found for reference `&AnonPipe` in the current scope
   --> library/std/src/os/windows/io/raw.rs:104:25
104 |         self.as_inner().as_raw_handle()
    |                         ^^^^^^^^^^^^^ method not found in `&AnonPipe`
    |
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `os::windows::io::raw::AsRawHandle` defines an item `as_raw_handle`, perhaps you need to implement it
   --> library/std/src/os/windows/io/raw.rs:22:1
22  | pub trait AsRawHandle {
    | ^^^^^^^^^^^^^^^^^^^^^


error[E0599]: no method named `as_raw_handle` found for reference `&AnonPipe` in the current scope
   --> library/std/src/os/windows/io/raw.rs:111:25
111 |         self.as_inner().as_raw_handle()
    |                         ^^^^^^^^^^^^^ method not found in `&AnonPipe`
    |
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `os::windows::io::raw::AsRawHandle` defines an item `as_raw_handle`, perhaps you need to implement it
   --> library/std/src/os/windows/io/raw.rs:22:1
22  | pub trait AsRawHandle {
    | ^^^^^^^^^^^^^^^^^^^^^

error[E0599]: no method named `into_raw_handle` found for struct `AnonPipe` in the current scope
---
19  | pub struct AnonPipe {
    | ------------------- method `into_raw_handle` not found for this struct
    |
    = help: items from traits can only be used if the trait is implemented and in scope
note: `os::windows::io::raw::IntoRawHandle` defines an item `into_raw_handle`, perhaps you need to implement it
   --> library/std/src/os/windows/io/raw.rs:78:1
78  | pub trait IntoRawHandle {
    | ^^^^^^^^^^^^^^^^^^^^^^^
help: there is a method `into_handle` with a similar name
    |
---
    |
210 |       unsafe fn from_raw_handle(raw_handle: RawHandle) -> io::PipeWriter {
    |                                                           -------------- expected `PipeWriter` because of return type
...
213 | /             io::PipeReader::from_inner(FromInner::from_inner(FromInner::from_inner(
214 | |                 OwnedHandle::from_raw_handle(handle),
    | |_______________^ expected `PipeWriter`, found `PipeReader`

error[E0599]: no method named `into_raw_handle` found for struct `AnonPipe` in the current scope
   --> library/std/src/os/windows/io/raw.rs:222:27
---
19  | pub struct AnonPipe {
    | ------------------- method `into_raw_handle` not found for this struct
    |
    = help: items from traits can only be used if the trait is implemented and in scope
note: `os::windows::io::raw::IntoRawHandle` defines an item `into_raw_handle`, perhaps you need to implement it
   --> library/std/src/os/windows/io/raw.rs:78:1
78  | pub trait IntoRawHandle {
    | ^^^^^^^^^^^^^^^^^^^^^^^
help: there is a method `into_handle` with a similar name
    |
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12975205954/job/36185990612) [(plain)](https://github.com/rust-lang/rust/commit/0778f24f7669d91038867426d5ce83163cd41515/checks/36185990612/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   --> library/std/src/io/pipe.rs:78:5
    |
76  | #[derive(Debug)]
    |          ----- in this derive macro expansion
77  | pub struct PipeReader {
78  |     inner: AnonPipe,
    |     ^^^^^^^^^^^^^^^ `AnonPipe` cannot be formatted using `{:?}`
   ::: /checkout/library/core/src/fmt/mod.rs:910:5
    |
910 |     pub macro Debug($item:item) {
    |     --------------- in this expansion of `#[derive(Debug)]`
    |     --------------- in this expansion of `#[derive(Debug)]`
    |
    = help: the trait `core::fmt::Debug` is not implemented for `AnonPipe`
    = note: add `#[derive(Debug)]` to `AnonPipe` or manually `impl core::fmt::Debug for AnonPipe`
help: consider annotating `AnonPipe` with `#[derive(Debug)]`
   --> library/std/src/sys/pal/windows/pipe.rs:19:1
19  + #[derive(Debug)]
20  | pub struct AnonPipe {
    |


error[E0277]: `AnonPipe` doesn't implement `core::fmt::Debug`
   --> library/std/src/io/pipe.rs:85:5
    |
83  | #[derive(Debug)]
    |          ----- in this derive macro expansion
84  | pub struct PipeWriter {
85  |     inner: AnonPipe,
    |     ^^^^^^^^^^^^^^^ `AnonPipe` cannot be formatted using `{:?}`
   ::: /checkout/library/core/src/fmt/mod.rs:910:5
    |
910 |     pub macro Debug($item:item) {
    |     --------------- in this expansion of `#[derive(Debug)]`
    |     --------------- in this expansion of `#[derive(Debug)]`
    |
    = help: the trait `core::fmt::Debug` is not implemented for `AnonPipe`
    = note: add `#[derive(Debug)]` to `AnonPipe` or manually `impl core::fmt::Debug for AnonPipe`
help: consider annotating `AnonPipe` with `#[derive(Debug)]`
   --> library/std/src/sys/pal/windows/pipe.rs:19:1
19  + #[derive(Debug)]
20  | pub struct AnonPipe {
    |


error[E0599]: the method `as_handle` exists for reference `&AnonPipe`, but its trait bounds were not satisfied
   --> library/std/src/os/windows/io/handle.rs:552:25
552 |         self.as_inner().as_handle()
    |                         ^^^^^^^^^
    |
   ::: library/std/src/sys/pal/windows/pipe.rs:19:1
   ::: library/std/src/sys/pal/windows/pipe.rs:19:1
    |
19  | pub struct AnonPipe {
    | ------------------- doesn't satisfy `AnonPipe: os::windows::io::handle::AsHandle`
note: trait bound `AnonPipe: os::windows::io::handle::AsHandle` was not satisfied
note: trait bound `AnonPipe: os::windows::io::handle::AsHandle` was not satisfied
   --> library/std/src/os/windows/io/handle.rs:445:9
    |
445 | impl<T: AsHandle + ?Sized> AsHandle for &T {
    |         |
    |         unsatisfied trait bound introduced here
note: the trait `os::windows::io::handle::AsHandle` must be implemented
note: the trait `os::windows::io::handle::AsHandle` must be implemented
   --> library/std/src/os/windows/io/handle.rs:426:1
426 | pub trait AsHandle {
    | ^^^^^^^^^^^^^^^^^^
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `os::windows::io::handle::AsHandle` defines an item `as_handle`, perhaps you need to implement it
   --> library/std/src/os/windows/io/handle.rs:426:1
426 | pub trait AsHandle {
    | ^^^^^^^^^^^^^^^^^^


error[E0599]: the method `as_handle` exists for reference `&AnonPipe`, but its trait bounds were not satisfied
   --> library/std/src/os/windows/io/handle.rs:571:25
571 |         self.as_inner().as_handle()
    |                         ^^^^^^^^^
    |
   ::: library/std/src/sys/pal/windows/pipe.rs:19:1
   ::: library/std/src/sys/pal/windows/pipe.rs:19:1
    |
19  | pub struct AnonPipe {
    | ------------------- doesn't satisfy `AnonPipe: os::windows::io::handle::AsHandle`
note: trait bound `AnonPipe: os::windows::io::handle::AsHandle` was not satisfied
note: trait bound `AnonPipe: os::windows::io::handle::AsHandle` was not satisfied
   --> library/std/src/os/windows/io/handle.rs:445:9
    |
445 | impl<T: AsHandle + ?Sized> AsHandle for &T {
    |         |
    |         unsatisfied trait bound introduced here
note: the trait `os::windows::io::handle::AsHandle` must be implemented
note: the trait `os::windows::io::handle::AsHandle` must be implemented
   --> library/std/src/os/windows/io/handle.rs:426:1
426 | pub trait AsHandle {
    | ^^^^^^^^^^^^^^^^^^
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `os::windows::io::handle::AsHandle` defines an item `as_handle`, perhaps you need to implement it
   --> library/std/src/os/windows/io/handle.rs:426:1
426 | pub trait AsHandle {
    | ^^^^^^^^^^^^^^^^^^


error[E0599]: no method named `as_raw_handle` found for reference `&AnonPipe` in the current scope
   --> library/std/src/os/windows/io/raw.rs:104:25
104 |         self.as_inner().as_raw_handle()
    |                         ^^^^^^^^^^^^^ method not found in `&AnonPipe`
    |
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `os::windows::io::raw::AsRawHandle` defines an item `as_raw_handle`, perhaps you need to implement it
   --> library/std/src/os/windows/io/raw.rs:22:1
22  | pub trait AsRawHandle {
    | ^^^^^^^^^^^^^^^^^^^^^


error[E0599]: no method named `as_raw_handle` found for reference `&AnonPipe` in the current scope
   --> library/std/src/os/windows/io/raw.rs:111:25
111 |         self.as_inner().as_raw_handle()
    |                         ^^^^^^^^^^^^^ method not found in `&AnonPipe`
    |
    = help: items from traits can only be used if the trait is implemented and in scope
    = help: items from traits can only be used if the trait is implemented and in scope
note: `os::windows::io::raw::AsRawHandle` defines an item `as_raw_handle`, perhaps you need to implement it
   --> library/std/src/os/windows/io/raw.rs:22:1
22  | pub trait AsRawHandle {
    | ^^^^^^^^^^^^^^^^^^^^^

error[E0599]: no method named `into_raw_handle` found for struct `AnonPipe` in the current scope
---
19  | pub struct AnonPipe {
    | ------------------- method `into_raw_handle` not found for this struct
    |
    = help: items from traits can only be used if the trait is implemented and in scope
note: `os::windows::io::raw::IntoRawHandle` defines an item `into_raw_handle`, perhaps you need to implement it
   --> library/std/src/os/windows/io/raw.rs:78:1
78  | pub trait IntoRawHandle {
    | ^^^^^^^^^^^^^^^^^^^^^^^
help: there is a method `into_handle` with a similar name
    |
---
    |
210 |       unsafe fn from_raw_handle(raw_handle: RawHandle) -> io::PipeWriter {
    |                                                           -------------- expected `PipeWriter` because of return type
...
213 | /             io::PipeReader::from_inner(FromInner::from_inner(FromInner::from_inner(
214 | |                 OwnedHandle::from_raw_handle(handle),
    | |_______________^ expected `PipeWriter`, found `PipeReader`

error[E0599]: no method named `into_raw_handle` found for struct `AnonPipe` in the current scope
   --> library/std/src/os/windows/io/raw.rs:222:27
---
19  | pub struct AnonPipe {
    | ------------------- method `into_raw_handle` not found for this struct
    |
    = help: items from traits can only be used if the trait is implemented and in scope
note: `os::windows::io::raw::IntoRawHandle` defines an item `into_raw_handle`, perhaps you need to implement it
   --> library/std/src/os/windows/io/raw.rs:78:1
78  | pub trait IntoRawHandle {
    | ^^^^^^^^^^^^^^^^^^^^^^^
help: there is a method `into_handle` with a similar name
    |
```

</details>

NobodyXu commented:
Hi @tbu- If you are busy and don't have time to continue this, I can open separate PRs to stablise annoynous_pipe and do the refactoring

nyabinary commented:
> Hi @tbu- If you are busy and don't have time to continue this, I can open separate PRs to stablise annoynous_pipe and do the refactoring

Prob a good idea :P

bors commented:
:umbrella: The latest upstream changes (presumably #138448) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

NobodyXu commented:
No worries, we all have our own lives and just doing this over spare time
